Orange/data/table/base.py

```diff
@@ -0,0 +1,1242 @@
+        elif isinstance(weight, pd.Series):  # not only SeriesBase
+            # drop everything but the values to uncomplicate things
+            if weight.isnull().any():
+                raise ValueError("Weights cannot be nan. ")
+            new_weights = weight.values
+        elif isinstance(weight, pd.Categorical):
+            new_weights = list(weight)
+        else:
+            raise TypeError("Expected one of [Number, str, Sequence, SeriesBase].")
+
+        # final check that we're actually adding numbers as weights
+        if not (isinstance(new_weights, Number) or all(isinstance(w, Number) for w in new_weights)):
+            raise ValueError("All weight values must be numbers.")
+        self[self._WEIGHTS_COLUMN] = new_weights
+
+    @property
+    def has_weights(self):
+        """Check if this Table has any weights.
+
+        Weight presence is determined by all weights not being identical.
+
+        Returns
+        -------
+        bool
+            True if the weights are not all identical, False otherwise.
+        """
+        return len(self) and len(self[self._WEIGHTS_COLUMN].unique()) != 1
+
+    @classmethod
+    def _new_id(cls, num=1, force_list=False):
+        """Generate new application-wide globally-unique numbers.
+
+        Parameters
+        ----------
+        num : int, optional, default 1
+            The number of new IDs to generate.
+        force_list : bool, optional, default False
+            Whether to always force the result to be a list,
+            even if only one number is generated.
+
+        Returns
+        -------
+        int or list[int]
+            New globally-unique identifiers.
+            If generating a single number and not forcing into a list, output an integer.
+            Otherwise output a list of integers of len(num).
+        """
+        with cls._next_instance_lock:
+            out = np.arange(cls._next_instance_id, cls._next_instance_id + num)
+            cls._next_instance_id += num
+            return out[0] if num == 1 and not force_list else out
+
+    def _transform_discrete_values_to_descriptors(self):
+        """Transform discrete variables given in index form into descriptors.
+
+        The conversion is performed in-place.
+        """
+        for var in chain(self.domain.variables, self.domain.metas):
+            if isinstance(var, DiscreteVariable):
+                # only transform the values if all of them appear to be integers and
+                # could act as a variable value index
+                # otherwise we're dealing with numeric discretes
+                is_values = self[var.name].apply(lambda v: isinstance(v, Number) and
+                                                           (isinstance(v, int) or v.is_integer()) and
+                                                           v < len(var.values)).all()
+                if is_values:
+                    self[var.name] = self[var.name].apply(lambda v: var.values[int(v)])
```

Comment:
These vectorized operations on pure numpy will always be 1000x times faster.
```py
values = self[var.name].values
try:
    ints = values.astype(np.uint16)
except ValueError:
    continue
# ...
if (ints == values).all() and ints.max() < len(var.values):
    self[var.name] = np.asanyarray(var.values)[ints]
```
Aren't they just as clear?